### PR TITLE
Use a Symbol to tag every ReactElement

### LINKED ISSUE #3473: How Much XSS Vulnerability Protection is React Responsible For?
There was a security hack that mentions React here: http://danlec.com/blog/xss-via-a-spoofed-react-element

Ultimately this is a server-side bug and NOT a bug in React itself. This issue is about figuring out if there is something we can do to mitigate issues when you have a JSON parsing bug or some server-side issue.

## isValidElement

React is designed to work with plain objects as input, and in fact, we're even getting rid of the `_isReactElement` as a way to verify this. We'll allow any JSON object. IMO, there is no problem with the verification here.

All string values are sanitized before inserted into the DOM (except for CSS styles which is a known wontfix issue for IE8).

In earlier versions we used `instanceof` checks but that didn't work well with multiple Reacts, it makes it difficult to optimize (inline objects are much faster) and couples JSX permanently to React, which we would like to avoid.

## dangerouslySetInnerHTML

One possible solution is to disable this feature and require it to be used imperatively (`React.findDOMNode(ref).innerHTML = ''`) which makes for worse performance at insertion time.

However, I don't believe this is the only bad thing once you can insert arbitrary HTML tags. It is certainly the easiest way to gain access to XSS though. You can also insert arbitrary Web Components which could expose data. You can render form elements that can potentially pass data.

## What else can we do?

Ultimately this is an issue where `<div>{userData}</div>` seems like a valid use case, but if your userData is compromised, it becomes dangerous.

Should React be responsible for protecting itself against arbitrary JSON as children?

### PR DESCRIPTION
**Pull request by sebmarkbage**
Fixes #3473

I tag each React element with `$$typeof: Symbol.for('react.element')`. We need
this to be able to safely distinguish these from plain objects that might have
come from user provided JSON.

The idiomatic JavaScript way of tagging an object is for it to inherent some
prototype and then use `instanceof` to test for it.

However, this has limitations since it doesn't work with value types which
require `typeof` checks. They also don't work across realms. Which is why there
are alternative tag checks like `Array.isArray` or the `toStringTag`. Another
problem is that different instances of React that might have been created not knowing about eachother. npm tends to make this kind of problem occur a lot.

Additionally, it is our hope that ReactElement will one day be specified in
terms of a "Value Type" style record instead of a plain Object.

This Value Types proposal by @nikomatsakis is currently on hold but does satisfy all these requirements:

https://github.com/nikomatsakis/typed-objects-explainer/blob/master/valuetypes.md#the-typeof-operator

Additionally, there is already a system for coordinating tags across module
systems and even realms in ES6. Namely using `Symbol.for`. (thx @sebmck)

Currently these objects are not able to transfer between Workers but there is
nothing preventing that from being possible in the future. You could imagine
even `Symbol.for` working across Worker boundaries. You could also build a
system that coordinates Symbols and Value Types from server to client or through
serialized forms. That's beyond the scope of React itself, and if it was built
it seems like it would belong with the `Symbol` system. A system could override
the `Symbol.for('react.element')` to return a plain yet
cryptographically random or unique number. That would allow ReactElements to
pass through JSON without risking the XSS issue.

The fallback solution is a plain well-known number. This makes it unsafe with
regard to the XSS issue described in #3473. We could have used a much more
convoluted solution to protect against JSON specifically but that would require
some kind of significant coordination, or change the check to do a
`typeof element.$$typeof === 'function'` check which would not make it unique to
React. It seems cleaner to just use a fixed number since the protection is just
a secondary layer anyway. I'm not sure if this is the right tradeoff.

In short, if you want the XSS protection, use a proper Symbol polyfill.

Finally, the reason for calling it `$$typeof` is to avoid confusion with `.type`
and the use case is to add a tag that the `typeof` operator would refer to.
I would use `@@typeof` but that seems to deopt in JSC. I also don't use
`__typeof` because this is more than a framework private. It should really be
part of the polyfilling layer.

Test Plan:

`examples/basic/index.html` works in Chrome and Firefox which both have native Symbols - and in Safari which doesn't.

`React.createElement('div').$$typeof // Symbol(react.element)`.

### CODE REVIEW

File: src/isomorphic/classic/element/ReactElement.js [hunk 1]
```diff
@@ -83,8 +88,25 @@ var ReactElement = function(type, key, ref, self, source, owner, props) {
...
+      // Two elements created in two different places should be considered
+      // equal for testing purposes and therefore we hide it from enumeration.
+      Object.defineProperty(element, '_source', {
+        configurable: false,
+        enumerable: false,
+        writable: false,
+        value: source,
+      });
     } else {
-      this._store.validated = false;
```
Comment (sebmarkbage) [inline src/isomorphic/classic/element/ReactElement.js]: Fixed this bug in older browsers.

File: src/isomorphic/classic/element/ReactElement.js [hunk 2]
```diff
@@ -52,17 +57,17 @@ if (__DEV__) {
...
   var element = {
+    // This tag allow us to uniquely identify this as a React Element
+    $$typeof: TYPE_SYMBOL,
+
     // Built-in properties that belong on the element
     type: type,
     key: key,
     ref: ref,
-    self: self,
-    source: source,
```
Comment (sebmarkbage) [inline src/isomorphic/classic/element/ReactElement.js]: These weren't intended for production use, and definitely not public. So I'm moving this to an underscore, `__DEV__` and making them non-enumerable. Otherwise jest's equality helpers won't work on elements created in two different source locations.

File: src/isomorphic/classic/element/ReactElement.js [hunk 3]
```diff
@@ -83,8 +88,25 @@ var ReactElement = function(type, key, ref, self, source, owner, props) {
...
+        configurable: false,
+        enumerable: false,
+        writable: false,
+        value: source,
+      });
     } else {
-      this._store.validated = false;
+      element._store.validated = false;
+      element._self = self;
+      element._source = source;
```
Comment (sebmarkbage) [inline src/isomorphic/classic/element/ReactElement.js]: I used a single underscore here for consistency with `_owner` and `_store` while the properties in `config` use the double underscore.

File: src/isomorphic/classic/element/ReactElement.js [hunk 4]
```diff
@@ -15,6 +15,11 @@ var ReactCurrentOwner = require('ReactCurrentOwner');
 
 var assign = require('Object.assign');
 
+// The Symbol used to tag the ReactElement type. If there is no native Symbol
+// nor polyfill, then a plain number is used for performance.
+var TYPE_SYMBOL = (typeof Symbol === 'function' && Symbol.for &&
+                  Symbol.for('react.element')) || 0xEAC7;
```
Comment (spicyj) [inline src/isomorphic/classic/element/ReactElement.js]: nit: `0xeac7` looks more like React ;)

Comment (spicyj) [inline src/isomorphic/classic/element/ReactElement.js]: Why `react.element` not `ReactElement` as we normally style it?

File: src/isomorphic/classic/element/__tests__/ReactElement-test.js [hunk 5]
```diff
@@ -24,6 +24,10 @@ describe('ReactElement', function() {
   beforeEach(function() {
     require('mock-modules').dumpCache();
 
+    // Delete the native Symbol if we have one to ensure we test the
+    // unpolyfilled environment.
+    delete global.Symbol;
```
Comment (spicyj) [inline src/isomorphic/classic/element/__tests__/ReactElement-test.js]: I guess this is configurable?

File: src/isomorphic/classic/element/ReactElement.js (hunk 4 above)
Comment (sebmarkbage) [inline src/isomorphic/classic/element/ReactElement.js]: Just because MDN recommended it: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/for

We could also do `react/element` to make it npm-like.

File: src/isomorphic/classic/element/__tests__/ReactElement-test.js (hunk 5 above)
Comment (sebmarkbage) [inline src/isomorphic/classic/element/__tests__/ReactElement-test.js]: What do you mean?

Comment (spicyj) [inline src/isomorphic/classic/element/__tests__/ReactElement-test.js]: Sorry:

I was under the impression that lots of globals aren't configurable and can't be deleted, but maybe that's not true at all. Evidently this one can be, at least.

Comment (sebmarkbage) [inline src/isomorphic/classic/element/__tests__/ReactElement-test.js]: Ah, yes. I was also not sure so I tested. I think it is generally just `window` and `document` and a few more that can't be deleted.

**Comment by glenjamin** [comment]
This is a great solution to the XSS issue! Just need to make sure no-one adds a way to run symbol.for via JSON parsing :)

File: src/isomorphic/classic/element/__tests__/ReactElement-test.js (hunk 5 above)
Comment (cpojer) [inline src/isomorphic/classic/element/__tests__/ReactElement-test.js]: jsdom is a little weird. Most fields are defined using Object.defineProperty with a getter only. In order to safely overwrite it, you need to use Object.defineProperty and set it's value to null. For forward compatibility with future jsdom updates I recommend doing that here.

Comment (spicyj) [inline src/isomorphic/classic/element/__tests__/ReactElement-test.js]: Well Symbol isn't from jsdom, right? Could add an `expect(typeof Symbol).toBe('undefined');` right after to be safe.

Also: we should probably restore the old Symbol global after this test.

**Comment by trueadm** [comment]
> works in Chrome and Firefox which both have native Symbols - and in Safari which doesn't.

I do believe this is very soon to change with the latest Safari for iOS 9 and El Capitan (16th September and 30th September according to yesterday's Apple event). :)

File: src/isomorphic/classic/element/__tests__/ReactElement-test.js (hunk 5 above)
Comment (cpojer) [inline src/isomorphic/classic/element/__tests__/ReactElement-test.js]: ohhh.. right, it is from the engine. Let's see if this test will fail with newer versions of node! Don't you still use node 0.10 for jest right now? If yes, that doesn't have Symbol, does it?

Comment (sebmck) [inline src/isomorphic/classic/element/__tests__/ReactElement-test.js]: Do you think it's worth requiring a symbol polyfill similar to how an ES5 one is required? Seems kinda dangerous to allow older browsers to be vulnerable.

Comment (chicoxyzzy) [inline src/isomorphic/classic/element/__tests__/ReactElement-test.js]: what @sebmck said

Comment (sebmarkbage) [inline src/isomorphic/classic/element/__tests__/ReactElement-test.js]: See the description in the PR summary. I don't know if it's the right tradeoff. This is a secondary layer of security so just because this hole doesn't exist doesn't mean it's exploitable. E.g. it has been in React since 0.13.

Comment (sebmarkbage) [inline src/isomorphic/classic/element/__tests__/ReactElement-test.js]: We have evaluated requiring various ES6 polyfills but decided against it in the past. It was too much of a hassle for people and didn't work well with node especially.

File: src/isomorphic/classic/element/ReactElement.js [hunk 6]
```diff
@@ -15,6 +15,11 @@ var ReactCurrentOwner = require('ReactCurrentOwner');
 
 var assign = require('Object.assign');
 
+// The Symbol used to tag the ReactElement type. If there is no native Symbol
+// nor polyfill, then a plain number is used for performance.
+var TYPE_SYMBOL = (typeof Symbol === 'function' && Symbol.for &&
+                  Symbol.for('react.element')) || 0xeac7;
```
Comment (gaearon) [inline src/isomorphic/classic/element/ReactElement.js]: 0xeac7 0xocc5!

File: src/isomorphic/classic/element/__tests__/ReactElement-test.js [hunk 7]
```diff
@@ -190,6 +194,10 @@ describe('ReactElement', function() {
     expect(React.isValidElement('string')).toEqual(false);
     expect(React.isValidElement(React.DOM.div)).toEqual(false);
     expect(React.isValidElement(Component)).toEqual(false);
+    expect(React.isValidElement({ type: 'div', props: {} })).toEqual(false);
+
+    var jsonElement = JSON.stringify(React.createElement('div'));
+    expect(React.isValidElement(JSON.parse(jsonElement))).toBe(true);
```
Comment (zpao) [inline src/isomorphic/classic/element/__tests__/ReactElement-test.js]: Is this supposed to work? Seems like this is the case we explicitly don't want to work (in the ideal case anyway).

I was giving node v4 a try where `Symbol` exists and this fails because `$$typeof` gets removed when stringifying.

Comment (sebmarkbage) [inline src/isomorphic/classic/element/__tests__/ReactElement-test.js]: That's why we try to delete global.Symbol to test the case where it doesn't exist but perhaps we need to shadow it rather than delete it. Didn't try Node 4.

Comment (zpao) [inline src/isomorphic/classic/element/__tests__/ReactElement-test.js]: Ok, just so I know, all of these tests here are explicitly for an environment where we don't have a native `Symbol`. In that case we expect `React.isValidElement(JSON.parse(JSON.stringify(element)))` to be `true`.

*BUT* if we do have a native `Symbol` (and weren't deleting it properly), then we would not expect this test to pass?

Comment (spicyj) [inline src/isomorphic/classic/element/__tests__/ReactElement-test.js]: That's right. See the next test case "identifies elements, but not JSON, if Symbols are supported" which tests this.

File: src/isomorphic/classic/element/ReactElement.js (hunk 6 above)
Comment (RReverser) [inline src/isomorphic/classic/element/ReactElement.js]: @gaearon `o` is not valid hex char :P

Comment (chicoxyzzy) [inline src/isomorphic/classic/element/ReactElement.js]: someone please tell me why 0xeac7? I can't sleep

Comment (zpao) [inline src/isomorphic/classic/element/ReactElement.js]: `0xeac7` sorta kinda looks like `React`

Comment (yiminghe) [inline src/isomorphic/classic/element/ReactElement.js]: why not use Math.random()

Comment (sebmarkbage) [inline src/isomorphic/classic/element/ReactElement.js]: Won't work between multiple instances of React and across realms.

Comment (STRML) [inline src/isomorphic/classic/element/ReactElement.js]: Wouldn't `Infinity` be a better magicnum? It can't be JSON-encoded, so we still get the XSS protection on browsers that don't support Symbols.

Comment (sebmarkbage) [inline src/isomorphic/classic/element/ReactElement.js]: @STRML That's a good point. I was hoping to avoid boxing. `NaN` or something like that would work too.

Comment (sebmarkbage) [inline src/isomorphic/classic/element/ReactElement.js]: It doesn't cover postMessage though.

Comment (STRML) [inline src/isomorphic/classic/element/ReactElement.js]: We *want* `postMessage` to work, right? In my testing, it actually does, since structured cloning supports much more advanced serialization than JSON.stringify().

This is confirmed in the w3c tests: https://github.com/w3c/web-platform-tests/blob/master/workers/interfaces/DedicatedWorkerGlobalScope/postMessage/structured-clone-message.html#L32

Comment (sebmarkbage) [inline src/isomorphic/classic/element/ReactElement.js]: We don't want it to work cross domain by default. That's still a security risk.

However, we do want it to work if you're able to coordinate the Symbol across the worker boundaries.

Comment (STRML) [inline src/isomorphic/classic/element/ReactElement.js]: I would argue that if you're deliberately catching cross-origin postMessage and inserting objects directly into your views without validation, you're asking for it.

This would be nice for webworkers though, as Symbols don't transfer via structured cloning (so you'd have to explicitly catch elements and re-tags them).

Comment (sebmarkbage) [inline src/isomorphic/classic/element/ReactElement.js]: See #3473 for more context.

It is too easy to expect a string and not realize it might be an object:

```js
loadString(function(stringData) {
  React.render(<div>Content: {stringData}</div>, container);
});
```
